core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java

```diff
@@ -879,8 +879,19 @@ private boolean checkConsistencyInternal(Inode inode, AlluxioURI path)
       return ufs.isDirectory(ufsPath);
     } else {
       InodeFile file = (InodeFile) inode;
-      return ufs.isFile(ufsPath)
-          && ufs.getFileStatus(ufsPath).getContentLength() == file.getLength();
+      if (ufs.isFile(ufsPath)) {
+        UfsFileStatus ufsFileStatus = ufs.getFileStatus(ufsPath);
+        if (ufs.getFileStatus(ufsPath).getContentLength() != file.getLength()) {
+          return false;
+        } else if (hasAnyBlockCached(mBlockMaster.getBlockInfoList(file.getBlockIds()))
+            && (ufsFileStatus.getLastModifiedTime() > file.getLastModificationTimeMs())) {
```

Comment:
Just sharing my thoughts :)
I think the best way to guarantee the consistency is to use a file integrity hash function like md5 or sha but this will probably use a lot for resources to read the files(which can be very large) and compute the hash.

If we still want to use the timestamps in consistency check, we could store the underlying filesystem timestamp separately from Alluxio's timestamp when the file is created or persisted and then use this value for consistency check. Additionally, I think comparing Alluxio's timestamp with the underlying filesystem timestamp can be very tricky since each filesystem has a different timestamp granularity and also files will usually be persisted at a timestamp that is different from the one when it is written in memory.